### PR TITLE
Allow assigning `null` to `Nullable`

### PR DESCRIPTION
**Pull request by GrimMaple**

### CODE REVIEW

File: std/typecons.d [hunk 1]
```diff
@@ -3059,25 +3061,31 @@ struct Nullable(T)
...
 
     /**
      * If this `Nullable` wraps a type that already has a null value
      * (such as a pointer), then assigning the null value to this
-     * `Nullable` is no different than assigning any other value of
-     * type `T`, and the resulting code will look very strange. It
-     * is strongly recommended that this be avoided by instead using
-     * the version of `Nullable` that takes an additional `nullValue`
-     * template argument.
+     * `Nullable` is no different than assigning $(REF nullify) ing it.
```
Comment (Herringway) [inline std/typecons.d:3076]: This wording seems pretty awkward. Did you mean to remove the word "assigning" as well?

File: std/typecons.d [hunk 2]
```diff
@@ -3059,25 +3061,31 @@ struct Nullable(T)
         _isNull = false;
         return this;
     }
+    /// ditto
+    Nullable opAssign()(typeof(null) value)
+    {
+        if(value !is null)
```
Comment (ljmf00) [inline std/typecons.d:3067]: Why is this here? Is it supposed to ever fall here?

Comment (GrimMaple) [reply std/typecons.d:3067]: I don't know what `typeof(null)` is on different platforms. I _assume_ it's going to be just for `null`, but there might be some obscure case later on. Though, I do think it's better to replace this entire `if` with `assert(value is null)`. Thank you!

File: std/typecons.d (hunk 1 above)
Comment (GrimMaple) [reply std/typecons.d:3076]: Yea, I seem to have forgotten to remove "assigning". Thank you! :)

**Comment by dkorpel** [comment]
Assigning `null` is different than setting the Nullable state to `null`, see: [The problem with Nullable: nullable(null).isNull is false.](https://forum.dlang.org/post/geozgznssrwsahjwaomf@forum.dlang.org)

I'm not sure this is a good idea. @atilaneves

File: std/typecons.d (hunk 2 above)
Comment (SixthDot) [reply std/typecons.d:3067]: No, `typeof(null)` converts to all pointers type and class types but nothing converts explictly to `typeof(null)`. The only value of `typeof(null)` is null itself.

**Comment by pbackus** [comment]
I think the real issue here is the name `Nullable`.

If it were called `Optional` or `Maybe`, nobody would expect assigning `null` to work like this in the first place, and nobody would be surprised to find out that it doesn't.

Unfortunately, we must keep the name `Nullable` for compatibility, but perhaps we could rename it and make `Nullable` an `alias`?

**Comment by GrimMaple** [comment]
Seeing how this PR isn't a particular hit with the D community, there is another solution I can propose: a compile-time error when trying to assign `null` to nullable saying something like "Use `nullify` instead"

@pbackus just renaming it to optional wouldn't be enough -- we would need to invent `nullopt` or something

File: std/typecons.d [hunk 3]
```diff
@@ -3059,25 +3061,34 @@ struct Nullable(T)
         _isNull = false;
         return this;
     }
+    /// ditto
+    Nullable opAssign()(typeof(null) value)
+    {
+        assert(value is null);
```
Comment (atilaneves) [inline std/typecons.d:3067]: Shouldn't this be a contract instead?

File: std/typecons.d [hunk 4]
```diff
@@ -3059,25 +3061,34 @@ struct Nullable(T)
...
         assert(npi.isNull);
 
-        //Passes?!
-        npi = null;
-        assert(!npi.isNull);
+        Nullable!int nint = Nullable!int(10);
+        nint = null;
+        assert(nint.isNull);
+
+        // Cannot assign other types with `null` value
```
Comment (atilaneves) [inline std/typecons.d:3089]: Why append to an existing test instead of creating a new one?

**Comment by atilaneves** [comment]
I'm not convinced of the utility of this either.

**Comment by GrimMaple** [comment]
I don't see a lot of movement on this, so I'm going to reask: would making a error on assigning `null` be more of an accepted solution?

**Comment by atilaneves** [comment]
> I don't see a lot of movement on this, so I'm going to reask: would making a error on assigning `null` be more of an accepted solution?

It's already a compile-time error, isn't it?

**Comment by GrimMaple** [comment]
> > I don't see a lot of movement on this, so I'm going to reask: would making a error on assigning `null` be more of an accepted solution?
> 
> It's already a compile-time error, isn't it?

Well it's not a particularly great compile-time error. I was suggesting explicitly erroring "Use `nullify` instead of assigning null"

**Comment by atilaneves** [comment]
> > > I don't see a lot of movement on this, so I'm going to reask: would making a error on assigning `null` be more of an accepted solution?
> > 
> > 
> > It's already a compile-time error, isn't it?
> 
> Well it's not a particularly great compile-time error. I was suggesting explicitly erroring "Use `nullify` instead of assigning null"

That sounds like a good idea.

**Comment by dkorpel** [comment]
If the `null` is a valid value, (e.g. Nullable!string), you should still be able to assign that. Otherwise (e.g. `Nullable!int`) it can give an error.

**Comment by LightBender** [comment]
We *really* like this one and want to get it merged, but it needs to handle nullable vs. non-nullable types. @GrimMaple any chance we could get you to look at this?

**Comment by GrimMaple** [comment]
> We _really_ like this one and want to get it merged, but it needs to handle nullable vs. non-nullable types. @GrimMaple any chance we could get you to look at this?

If you can come up with a community consensus on what the behavior should be, I can implement it no problem. I don't want to invest my time into (most likely, fruitlessly) trying to please everyone

**Comment by LightBender** [comment]
> If the `null` is a valid value, (e.g. `Nullable!string`), you should still be able to assign that. Otherwise (e.g. `Nullable!int`) it can give an error.

@dkorpel This is fundamentally illogical. The whole purpose of a Nullable is to make non-nullable types behave as if they are "nullable". Furthermore, calling `.nullify()` does not generate an error when called on a non-nullable base type meaning that this design would result in logically inconsistent behavior.

@GrimMaple I'm not entirely sure why this is even a debate. The only rational thing to do is allow setting null for all types since that is the whole point of a Nullable. The same logic would apply for `is null` checks.

If we want an `Optional` type that behaves like an Optional Type, then we can build that, but Optional Types are a fundamentally different beasts and are trying to accomplish different things than Nullable types. Trying to transmogrify Nullable's into Optional's is fundamentally bad design.

Nominally, Optional Types can be used to accomplish what the Nullable does. But A) we already have Nullable (backwards compat problem) and B) Optionals are supersets of Nullable's. Therefore by trying to make Nullable into an Optional is just creating a crippled version of both.

@pbackus @atilaneves Unless you have an extremely strong argument as to why Nullable needs to violate basic logic so that it can behave like a crippled Optional type, I will be requesting that @GrimMaple to move forward with this PR on the basis of allowing null assignment to all Nullable types regardless of underlying type. If we want Optional types, that will be a separate discussion.

**Comment by jmdavis** [comment]
I object to this change on the grounds that this change would make `Nullable` even more of a mess than it already is and that it could break existing code. With this change, code such as

```
void main()
{
    import std.typecons;

    Nullable!(int*) n;
    assert(n.isNull);
    assert(n = null);
    assert(!n.isNull);
}
```

would change behavior. And code that wants to detect whether a variable has a value as opposed to whether that value is `null` will care about this change.

`Nullable` has _never_ been about making a type actually `null`. If you look at the code, it never sets its value to `null` even if the type is actually nullable, and it's perfectly valid to assign it a `null` as a value (which does not make `isNull == true`) - which if anything, highlights part of the problem with the naming scheme here. `Nullable` has always been an "optional" type, and for someone coming from a C++ background (particularly when `Nullable` was introduced), naming it `Nullable` made sense, because there was no "optional" type in C++ until C++17, and talking about types being nullable is the obvious comparison, since you're effectively trying to make a type that isn't nullable be "nullable". But it was never about the value `null`, and `Nullable` was never coded that way. IIRC, Andrei was the one who named it.

Either way, naming it `Nullable` was clearly a mistake, because folks have come along for years and tried to make it interact with `null` regardless of whether the type itself was nullable, which causes problems for types which are actually nullable, since it does not treat `null` as a special value, and there is no way to have `is null` work with a user-defined type, because you can't override the `is` operator. So, having `Nullable` interact with `null` in the way that it does with actually nullable types simply isn't possible, and trying to treat `null` as special with `Nullable` will break any code that currently sets a `Nullable` to `null`, since while that's arguably not a great idea, `isNull` and `null` are distinct with nullable types, and in fact, it's not uncommon for folks to purposefully use `Nullable` with arrays to distinguish from `null` because of the issues with conflating `empty` and `null` in cases where you want to actually check whether there's a value or not rather than whether the array is empty. And for types in general, it's a great way to generically test whether a value has been set without worrying about whether the type is actually nullable, and in some of those cases, `null` will be considered a valid value by the non-generic code around it.

I totally think that we should use a different name for Phobos v3, since naming this type `Nullable` has simply caused confusion and problems, but please lets not change its current behavior any further. The fact that it was changed to be a range already has been causing issues, and changing its behavior yet further really does risk breaking existing code.

**Comment by LightBender** [comment]
<sigh>

<mega-sigh>

Ok.

After having @jmdavis explain this to me in excruciating detail. I will close it.

However, I want to make it crystal clear that this entire situation is a monumental failure in naming. A Nullable that really wants to be an Option type is guaranteed to confuse every programmer on the planet, including some of D's finest. @GrimMaple is no slouch. As such, Phobos 3 will not use the `Nullable` name for this functionality. I will link the Phobos 3 PR that adds `Option` to this for one posterity.

If people start whining when I come after them about sloppy naming, I'm going to use this as a prime example of why I am being such a zealot.

My apologies for the result @GrimMaple, if you want more details you can reach out to me or JMD on Discord.

**Comment by pbackus** [comment]
@LightBender For reference, there was some discussion on renaming `Nullable` for Phobos 3 here:

https://github.com/dlang/PhobosV3-Design/discussions/15#discussioncomment-8424375
